analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -627,3 +633,55 @@ public void Method(object o)
         }
     }
 }
+
+public class Conversion
+{
+    public void DownCast(object o)
+    {
+        ((string)o).ToString();                         // Noncompliant
+    }
+
+    public void UpCast(string s)
+    {
+        ((object)s).ToString();                         // Noncompliant
+    }
+
+    public void CastWithPropertyAccess(object o)
+    {
+        _ = ((Exception)o).Message;                     // Noncompliant
+    }
+
+    public void CastWithRedundantParentheses(object o)
+    {
+        (((string)o)).ToString();                       // Noncompliant
+    }
+
+    public void MultipleCasts(object o)
+    {
+        ((string)((object)((string)o))).ToString();     // Noncompliant
+    }
+
+    public void AsOperatorDownCast(object o)
+    {
+        (o as string).ToString();                       // Noncompliant
+    }
+
+    public void AsOperatorUpCast(string s)
+    {
+        (s as object).ToString();                       // Noncompliant
+    }
+
+    public void ForEachLoop(object[] arr)
+    {
+        foreach (object o in arr)                       // Noncompliant - the array is first cast to an IEnumerable, then the GetEnumerator method is invoked on it
+        {
+        }
+    }
+
+    public void ForEachLoopWithCast(object[] arr)
+    {
+        foreach (object o in (IEnumerable<object>)arr)  // Noncompliant
+        {
+        }
+    }
+}
```

Comment:
I would also add the following unit test, which was not working before the change from `Parent` to `Ancestors()`, and works now:
```cs
public void Parentheses(object[] arr)
{
    foreach (object o in ((arr))) { } // Noncompliant
}
```

If you want to push it one level further:
```cs
public void Parentheses(object[] arr)
{
    foreach (object o in ((object[])(object)((arr)))) { } // Noncompliant
}
```
and similar.